### PR TITLE
fix(@libp2p/webtransport): be more thorough about closing sessions

### PR DESCRIPTION
**Pull request by achingbrain**
Refactors session closing to happen in one function and call that function when the session has closed or failed to init.

Doesn't quite solve the "Too many pending WebTransport Sessions" problem but does slow it down a little bit.

Refs: #1896

### CODE REVIEW

File: packages/transport-webtransport/src/index.ts [hunk 1]
```diff
@@ -52,60 +64,122 @@ class WebTransportTransport implements Transport {
...
+            reason: err.message
+          })
+        },
+        remoteAddr: ma,
+        timeline: {
+          open: Date.now()
+        },
+        // This connection is never used directly since webtransport supports native streams.
+        ...inertDuplex()
+      }
```
Comment (SgtPooki) [inline packages/transport-webtransport/src/index.ts:164]: if we make this a property on the class we can split out this dial function a bit

File: packages/transport-webtransport/src/index.ts [hunk 2]
```diff
@@ -52,60 +64,122 @@ class WebTransportTransport implements Transport {
...
+        .catch((err: Error) => {
+          log.error('error on remote wt session close', err)
+          maConn?.abort(err)
+        })
+        .finally(() => {
+          // if we never got as far as creating the maConn, just clean up the session
+          if (maConn == null) {
+            cleanUpWTSession()
+          }
+        })
```
Comment (SgtPooki) [inline packages/transport-webtransport/src/index.ts:128]: I believe this should cover the case where `await wt.ready` throws, but from my testing, I'm not confident the `.closed` catch always handles errors thrown from `wt.ready`.

**Review by SgtPooki — APPROVED** [review]
definitely an improvement

File: packages/transport-webtransport/src/index.ts [hunk 3]
```diff
@@ -52,60 +64,122 @@ class WebTransportTransport implements Transport {
...
     try {
+      let closed = false
+      const wt = new WebTransport(`${url}/.well-known/libp2p-webtransport?type=noise`, {
+        serverCertificateHashes: certhashes.map(certhash => ({
+          algorithm: 'sha-256',
+          value: certhash.digest
+        }))
+      })
+
+      cleanUpWTSession = (closeInfo?: WebTransportCloseInfo) => {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/transport-webtransport/src/index.ts:89]: It needed access to a lot of scoped variables, and it's passed to the stream muxer so it got a bit fiddly.

**Pull request merged**